File: content/_code-samples/send-xrp/send-xrp.js
// Example credentials
const wallet = xrpl.Wallet.fromSeed("sn3nxiW7v8KXzPzAqzyHXbSSKNuN9")

// Connect ---------------------------------------------------------------------
// const  xrpl = require('xrpl') // For Node.js. In browsers, use <script>.
const api = new xrpl.Client('wss://s.altnet.rippletest.net:51233')
api.connect()
api.on('connected', async () => {

  // Get credentials from the Testnet Faucet -----------------------------------
  const wallet = await api.generateFaucetWallet()

  console.log("Waiting until we have a validated starting sequence number...")
  // If you go too soon, the funding transaction might slip back a ledger and
  // then your starting Sequence number will be off. This is mostly relevant
  // when you want to use a Testnet account right after getting a reply from
  // the faucet.
  while (true) {
    try {
      await api.request({
        command: "account_info",
        account: wallet.classicAddress,
        ledger_index: "validated"
      })
      break
    } catch(e) {
      await new Promise(resolve => setTimeout(resolve, 1000))
    }
  }

  // Prepare transaction -------------------------------------------------------
  const prepared = await api.autofill({
    "TransactionType": "Payment",
    "Account": wallet.classicAddress,
    "Amount": xrpl.xrpToDrops("22"),
    "Destination": "rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe"
  })
  const max_ledger = prepared.LastLedgerSequence
  console.log("Prepared transaction instructions:", prepared)
  console.log("Transaction cost:", xrpl.dropsToXrp(prepared.Fee), "XRP")
  console.log("Transaction expires after ledger:", max_ledger)

  // Sign prepared instructions ------------------------------------------------
  const tx_blob = wallet.signTransaction(prepared)
  const txID = xrpl.computeSignedTransactionHash(tx_blob)
  console.log("Identifying hash:", txID)
  console.log("Signed blob:", tx_blob)

  // Submit signed blob --------------------------------------------------------
  // The earliest ledger a transaction could appear in is the first ledger
  // after the one that's already validated at the time it's *first* submitted.
  const min_ledger = (await api.getLedgerIndex()) + 1
  const result = await api.request({
    "command": "submit",
    "tx_blob": tx_blob
  })
  console.log("Tentative result code:", result.result.engine_result)
  console.log("Tentative result message:", result.result.engine_result_message)

  // Wait for validation -------------------------------------------------------
  let has_final_status = false
  api.request({
    "command": "subscribe",
    "accounts": [wallet.classicAddress]
  })
  api.connection.on("transaction", (event) => {
    if (event.transaction.hash == txID) {
      console.log("Transaction has executed!", event)
      has_final_status = true
    }
  })
  api.on('ledgerClosed', ledger => {
    if (ledger.ledger_index > max_ledger && !has_final_status) {
      console.log("Ledger version", ledger.ledger_index, "was validated.")
      console.log("If the transaction hasn't succeeded by now, it's expired")
      has_final_status = true
    }
  })

  // There are other ways to do this, but they're more complicated.
  // See https://xrpl.org/reliable-transaction-submission.html for details.
  while (!has_final_status) {
    await new Promise(resolve => setTimeout(resolve, 1000))
  }

  // Check transaction results -------------------------------------------------
  try {
    const tx = await api.request({
      command: "tx",
      transaction: txID,
      min_ledger: min_ledger,
      max_ledger: max_ledger
    })
    if (tx.result.validated) {
      console.log("This result is validated by consensus and final.")
    } else {
      console.log("This result is pending.")
    }
    console.log("Transaction result:", tx.result.meta.TransactionResult)

    if (typeof tx.result.meta.delivered_amount === "string" &&
        typeof tx.result.meta.delivered_amount !== "unavailable")
    console.log("Delivered:", xrpl.dropsToXrp(tx.result.meta.delivered_amount), "XRP")
  } catch(error) {
    console.log("Couldn't get transaction outcome:", error)
  }

}) // End of api.on.('connected')

File: content/_code-samples/send-xrp/SendXrp.java
// Example Credentials --------------------------------------------------------
WalletFactory walletFactory = DefaultWalletFactory.getInstance();
Wallet testWallet = walletFactory
  .fromSeed("sn3nxiW7v8KXzPzAqzyHXbSSKNuN9", true)
  .wallet();

// Get the Classic address from testWallet
Address classicAddress = testWallet.classicAddress();
System.out.println(classicAddress); // "rMCcNuTcajgw7YTgBy1sys3b89QqjUrMpH"

// Connect --------------------------------------------------------------------
HttpUrl rippledUrl = HttpUrl.get("https://s.altnet.rippletest.net:51234/");
XrplClient xrplClient = new XrplClient(rippledUrl);

// Prepare transaction --------------------------------------------------------
// Look up your Account Info
AccountInfoRequestParams requestParams = AccountInfoRequestParams.builder()
  .ledgerIndex(LedgerIndex.VALIDATED)
  .account(classicAddress)
  .build();
AccountInfoResult accountInfoResult = xrplClient.accountInfo(requestParams);
UnsignedInteger sequence = accountInfoResult.accountData().sequence();

// Request current fee information from rippled
FeeResult feeResult = xrplClient.fee();
XrpCurrencyAmount openLedgerFee = feeResult.drops().openLedgerFee();

// Get the latest validated ledger index
LedgerIndex validatedLedger = xrplClient.ledger(
    LedgerRequestParams.builder()
      .ledgerIndex(LedgerIndex.VALIDATED)
      .build()
  )
  .ledgerIndex()
  .orElseThrow(() -> new RuntimeException("LedgerIndex not available."));

// Workaround for https://github.com/XRPLF/xrpl4j/issues/84
UnsignedInteger lastLedgerSequence = UnsignedInteger.valueOf(
  validatedLedger.plus(UnsignedLong.valueOf(4)).unsignedLongValue().intValue()
);

// Construct a Payment
Payment payment = Payment.builder()
  .account(classicAddress)
  .amount(XrpCurrencyAmount.ofXrp(BigDecimal.ONE))
  .destination(Address.of("rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe"))
  .sequence(sequence)
  .fee(openLedgerFee)
  .signingPublicKey(testWallet.publicKey())
  .lastLedgerSequence(lastLedgerSequence)
  .build();
System.out.println("Constructed Payment: " + payment);

// Sign transaction -----------------------------------------------------------
// Construct a SignatureService to sign the Payment
PrivateKey privateKey = PrivateKey.fromBase16EncodedPrivateKey(
  testWallet.privateKey().get()
);
SignatureService signatureService = new SingleKeySignatureService(privateKey);

// Sign the Payment
SignedTransaction<Payment> signedPayment = signatureService.sign(
  KeyMetadata.EMPTY,
  payment
);
System.out.println("Signed Payment: " + signedPayment.signedTransaction());

// Submit transaction ---------------------------------------------------------
SubmitResult<Transaction> prelimResult = xrplClient.submit(signedPayment);
System.out.println(prelimResult);

// Wait for validation --------------------------------------------------------
boolean transactionValidated = false;
boolean transactionExpired = false;
while (!transactionValidated && !transactionExpired) {
  Thread.sleep(4 * 1000);
  LedgerIndex latestValidatedLedgerIndex = xrplClient.ledger(
      LedgerRequestParams.builder().ledgerIndex(LedgerIndex.VALIDATED).build()
    )
    .ledgerIndex()
    .orElseThrow(() ->
      new RuntimeException("Ledger response did not contain a LedgerIndex.")
    );

  TransactionResult<Payment> transactionResult = xrplClient.transaction(
    TransactionRequestParams.of(signedPayment.hash()),
    Payment.class
  );

  if (transactionResult.validated()) {
    System.out.println("Payment was validated with result code " +
      transactionResult.metadata().get().transactionResult());
    transactionValidated = true;
  } else {
    boolean lastLedgerSequenceHasPassed = FluentCompareTo.
      is(latestValidatedLedgerIndex.unsignedLongValue())
      .greaterThan(UnsignedLong.valueOf(lastLedgerSequence.intValue()));
    if (lastLedgerSequenceHasPassed) {
      System.out.println("LastLedgerSequence has passed. Last tx response: "
        transactionResult);
    );
      transactionExpired = true;
    } else {
      System.out.println("Payment not yet validated.");
    }
  }
}

// Check transaction results --------------------------------------------------
System.out.println(transactionResult);
System.out.println("Explorer link: https://testnet.xrpl.org/transactions/" +
    signedPayment.hash());
transactionResult.metadata().ifPresent(metadata -> {
  System.out.println("Result code: " + metadata.transactionResult());

  metadata.deliveredAmount().ifPresent(deliveredAmount ->
    System.out.println("XRP Delivered: " +
      ((XrpCurrencyAmount) deliveredAmount).toXrp())
  );
});
